### ISSUE TITLE
Create a repo and package for Jaeger.Thrift

### ISSUE THREAD

**Issue by grounded042** [Chatham/LetsTrace.Jaeger#2]
See the thread in #1 for more info on this.

We want to keep the generated IDL separate from the Jaeger client. We should create a repo for it as well as a NuGet package. The repo should use sub-modules to pull in the [jaeger-idl repo](https://github.com/jaegertracing/jaeger-idl/tree/master/thrift). We should have a script to run the generation. We should also ILRepack the thrift .Net library inside of the package until Thrift officially comes out with a NuGet package.

**Comment by mweinand** [comment]
@yurishkuro I don't have an opinion one way or the other on separate repository, but it makes sense to me to follow the pattern set by jaeger-client-java.

Is the goal to turn this (`LetsTrace.Jaeger`) repository into `jaeger-client-dotnet`?  Or will there be a separate `jaeger-client-dotnet`?  Would anyone have value referencing the `Jaeger.Thrift` project separate from `jaeger-client-dotnet`? Since it is basically just compiling outputs from an upstream source, I doubt `Jaeger.Thrift` is going to need to maintain its own Issues list or pull requests.

**Comment by grounded042** [comment]
Copying this here to continue the convo here. I said the following over in #1

> we certainly could put them in the same repo - I've seen different people do things different ways for C#. In the current workflow we have setup for the publishing of the NuGet we would always publish a NuGet package for LetsTrace.Jaeger and Jaeger.Thrift even if only one of them had changes. This is more of a build system limitation and I'll look into alternate systems besides Appveyor.

**Comment by mweinand** [comment]
Looks like the rest of Jaeger is using Travis: https://travis-ci.org/jaegertracing so it may make sense to consolidate there?

**Comment by grounded042** [comment]
We would like to be able to have this project become the official jaeger client for C#.

**Comment by yurishkuro** [comment]
> Is the goal to turn this (LetsTrace.Jaeger) repository into jaeger-client-dotnet? 

My suggestion was to execute a move & rename of this repo into jaegertracing/jaeger-client-csharp. I am not 100% clear on the different between `LetsTrace.Jaeger` and `LetsTrace`, would be great to merge everything into one repo before the move.

Also, we don't have to wait for "official" status before we do the move, it's more about releasing 1.0 and the namespace.

**Comment by mweinand** [comment]
Not sure if Travis supports only running a portion of the build (publishing `Jaeger.Thrift`) on a manual step or not (only when someone opts-in to it).

**Comment by grounded042** [comment]
It looks like the Java repo publishes everything each time. Though I'm not very familiar with how packages work in Java.

**Comment by mweinand** [comment]
Actually, if we don't increment the version number for `Jaeger.Thrift` we can likely just let NuGet.org respond with an overwrite failure (don't force it) and as long as that doesn't fail the rest of the build it is probably okay.

**Comment by grounded042** [comment]
> we don't have to wait for "official" status before we do the move, it's more about releasing 1.0 and the namespace.

Good to know. Hang on and I'll get back to you on this @yurishkuro

**Comment by yurishkuro** [comment]
yes, java repo versions and publishes all modules at the same time.

**Comment by mweinand** [comment]
@grounded042 to @yurishkuro 's point, how does the base `LetsTrace` project fit into the rest of this? Is that project meant to be a common wrapper around other things that might not be Jaeger?

**Comment by grounded042** [comment]
The original goal of LetsTrace was to have a main tracing lib with plugins for different tracing systems like Jaeger and Zipkin. However, it seems like the need / want is for a Jaeger C# client. Talking with my team we've been thinking that it might be better to merge them like @yurishkuro said.

**Comment by mweinand** [comment]
Forgive my ignorance, but are there features in the base LetsTrace project where we might just utilize the base OpenTracing components?  For instance, I'm not well versed in the additional features provided by https://github.com/Chatham/LetsTrace/blob/master/src/LetsTrace/ILetsTraceSpan.cs vs https://github.com/opentracing/opentracing-csharp/blob/master/src/OpenTracing/ISpan.cs

**Comment by mweinand** [comment]
Okay I think I've got a better feel for how everything works together.  OpenTracing is just interfaces, LetsTrace provides some base implementations, then LetsTrace.Jaeger is the specific Jaeger implementations that extend LetsTrace.

**Comment by yurishkuro** [comment]
From my brief review of LetsTrace alone, it's still very much repeats the overall structure of Jaeger clients. Yes, it can be used for other backends like Zipkin (so can other Jaeger clients, e.g. in Java or Go), but not for much else beyond that because it already enforces a certain data model and trace context semantics.

**Comment by grounded042** [comment]
We’d like to get the two repos merged and moved @yurishkuro. I’m thinking we should merge the two improvement branches into LetsTrace and LetsTrace.Jaeger and then merge those to repos together and do the rename and move.

**Comment by Falco20019** [comment]
The overall structure is quite similar to the java client.
`jaeger-core`: `LetsTrace` + parts of `LetsTrace.Jaeger`
`jaeger-thrift`: `Jaeger.Thrift` which is part of `LetsTrace.Jaeger`
`jaeger-zipkin`: `LetsTrace.Zipkin`

So while `jaeger-core` already has a reference on `jaeger-thrift` and includes the Jaeger protocol, `LetsTrace` also abstracted that away. So in case that you only want to use `Zipkin`, with `LetsTrace` you could exclude it (and all of the Thrift runtime and IDL code), while with Java `jaeger-core` it's always included.

I think it would be nice to go the route that `Grpc` went and have a `Jaeger.Core` NuGET package (which is basically `LetsTrace`) which can not run by itself since it's missing an `ISender` implementation. Then we have a `Jaeger` NuGET packages (which is `LetsTrace.Jaeger` with dependency on `Jaeger.Core`) and we have a `Jaeger.Zipkin` package (which is `LetsTrace.Zipkin` with dependency on `Jaeger.Core`).

**Comment by yurishkuro** [comment]
+1. Java project structure is not exactly clean.
